src_features/signMessageEIP712/ui_logic.c

```diff
@@ -0,0 +1,548 @@
+#ifdef HAVE_EIP712_FULL_SUPPORT
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include "ui_logic.h"
+#include "mem.h"
+#include "mem_utils.h"
+#include "os_io.h"
+#include "shared_context.h"
+#include "ethUtils.h"  // getEthDisplayableAddress
+#include "utils.h"     // uint256_to_decimal
+#include "common_712.h"
+#include "context_712.h"     // eip712_context_deinit
+#include "uint256.h"         // tostring256 && tostring256_signed
+#include "path.h"            // path_get_root_type
+#include "apdu_constants.h"  // APDU response codes
+#include "typed_data.h"
+#include "commands_712.h"
+#include "common_ui.h"
+
+static t_ui_context *ui_ctx = NULL;
+
+/**
+ * Checks on the UI context to determine if the next EIP 712 field should be shown
+ *
+ * @return whether the next field should be shown
+ */
+static bool ui_712_field_shown(void) {
+    bool ret = false;
+
+    if (ui_ctx->filtering_mode == EIP712_FILTERING_BASIC) {
+        if (N_storage.verbose_eip712 || (path_get_root_type() == ROOT_DOMAIN)) {
+            ret = true;
+        }
+    } else  // EIP712_FILTERING_FULL
+    {
+        if (ui_ctx->field_flags & UI_712_FIELD_SHOWN) {
+            ret = true;
+        }
+    }
+    return ret;
+}
+
+/**
+ * Set UI buffer
+ *
+ * @param[in] src source buffer
+ * @param[in] src_length source buffer size
+ * @param[in] dst destination buffer
+ * @param[in] dst_length destination buffer length
+ * @param[in] explicit_trunc if truncation should be explicitely shown
+ */
+static void ui_712_set_buf(const char *const src,
+                           size_t src_length,
+                           char *const dst,
+                           size_t dst_length,
+                           bool explicit_trunc) {
+    uint8_t cpy_length;
+
+    if (src_length < dst_length) {
+        cpy_length = src_length;
+    } else {
+        cpy_length = dst_length - 1;
+    }
+    memcpy(dst, src, cpy_length);
+    dst[cpy_length] = '\0';
+    if (explicit_trunc && (src_length > dst_length)) {
```

Comment:
>= non ?

Reply:
It wouldn't have created an out-of-band write, but indeed it could have missed adding the ... because of that 1 byte offset.
Fixed in [174d357](https://github.com/LedgerHQ/app-ethereum/pull/327/commits/174d35782f739d2da5beeec2006831fae79b4d30)